### ISSUE TITLE
Bug: [explicit-function-return-type] false negatives on inner function when using allowTypedFunctionExpressions option

### ISSUE THREAD

**Issue by yeonjuan** [typescript-eslint/typescript-eslint#8808]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&showAST=es&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBKAXIgbnGAE0QG8AoRRCBAZykQCMBDAJ0QF5FMOA%2BLjXqUQB6YYgBycRAFEWLOCwA0yJgBsaAU0RgNAcyaxcGipUQsNUECyTUwNOKo0A6VXF1oA5CjgeMAbhMAXwDAsjJbOkZWDn5BbnY%2BEhExWXk2JhpEDQAPAAcNaA1CJxNzS2sqWgdnV3cvVF8QvyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0YANqRE0aGOiQANJnDYcAQ3jwxAdwAqJRABMAYkJHimAUS7TkyV6gz44REMIAF8AXUMQkBCgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function foo(): void {
  const bar = () => () => { // No Error, false negative
    return console.log('foo');
  };
}

const bar = () => () => { // Error as expected.
  return console.log('foo');
};
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
  "@typescript-eslint/explicit-function-return-type": ["error", {
      "allowTypedFunctionExpressions": true
    }]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Two errors (line 2, line 7)

### Actual Result

One error (line 7)

### Additional Info

If the `allowTypedFunctionExpressions` option is false, this bug does not occur. IMO, [ancestorHasReturnType](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/explicit-function-return-type.ts#L215) seems to be buggy when traversing ancestor nodes: when encountering a certain node type, it should return false without further traversal.

[playground](https://typescript-eslint.io/play/#ts=5.4.3&showAST=es&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBKAXIgbnGAE0QG8AoRRCBAZykQCMBDAJ0QF5FMOA%2BLjXqUQB6YYgBycRAFEWLOCwA0yJgBsaAU0RgNAcyaxcGipUQsNUECyTUwNOKo0A6VXF1oA5CjgeMAbhMAXwDAsjJbOkZWDn5BbnY%2BEhExWXk2JhpEDQAPAAcNaA1CJxNzS2sqWgdnV3cvVF8QvyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0YANqRE0aGOiQANJnDYcAQ3jwxAdwAqJRABMAYkJHimAUS7TkyV6gz9TZERDCABfAF1DUJBQoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
```ts
function foo(): void {
  const bar = () => () => { // Error as expected
    return console.log('foo');
  };
}

const bar = () => () => { // Error as expected.
  return console.log('foo');
};
```
